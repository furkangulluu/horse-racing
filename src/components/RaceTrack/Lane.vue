<template>
  <div class="lane flex items-center" ref="laneRef" :data-lane="props.laneNo">
    <div
      class="horse overflow-hidden"
      :class="{ running: props.horse.run }"
      :style="{
        color: horse.color,
        'margin-left': props.horse.travelledDistance + 'px',
      }"
    >
      üêé
    </div>
  </div>
</template>

<script lang="ts" setup>
import { ref, PropType } from "vue";
import { IHorse } from "@/models/Horse";

const props = defineProps({
  horse: {
    type: Object as PropType<IHorse>,
    required: true,
  },
  laneNo: {
    type: Number,
    default: 0,
  },
});

const color = ref(props.horse.color);
</script>

<style lang="scss">
.lane {
  position: relative;
  height: 50px;
  max-width: 100vw;

  & .horse {
    position: relative;
    transition: all 0.15s ease-in-out;
    font: 50px Muybridge;

    &.running {
      animation: 1s linear infinite Gallop;
    }
  }

  &::before {
    position: absolute;
    content: attr(data-lane);
    top: 50%;
    transform: translateY(-50%);
    height: 43px;
    width: 15px;
    margin-left: -20px;
    background-color: rgb(26, 230, 43);
    color: var(--color-white);
    font-size: small;
    font-weight: bold;
    display: flex;
    align-items: center;
    justify-content: center;
    z-index: 2;
  }

  &::after {
    position: absolute;
    content: "";
    top: 50%;
    right: 20px;
    height: 90%;
    width: 10px;
    transform: translateY(-50%);
    z-index: 0;
    background-color: var(--color-white);
    background-size: 5px 5px;
    background-image: repeating-linear-gradient(
      45deg,
      #222226 0,
      #222226 0.5px,
      #ffffff 0,
      #ffffff 50%
    );
  }
}

@keyframes Gallop {
  from {
    font-variation-settings: "TIME" 0;
  }
  to {
    font-variation-settings: "TIME" 15;
  }
}
</style>
