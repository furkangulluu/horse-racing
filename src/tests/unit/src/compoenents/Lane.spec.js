import { describe, expect, test } from "vitest";
import { mount } from "@vue/test-utils";
import Lane from "@/components/RaceTrack/Lane.vue";

describe("Lane.vue", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(Lane, {
      props: {
        horse: {
          id: "adb367f2-73be-4699-a9c2-200ce3006a08",
          name: "Horse 1",
          condition: 70,
          color: "#000000",
          travelledDistance: 0,
          speed: 0,
          finish: false,
          run: false,
        },
        laneNo: 1,
      },
    });
  });

  test("is correctly rendered", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("renders props.laneNo when passed", () => {
    expect(wrapper.vm.laneNo).toEqual(1);
    expect(wrapper.attributes("data-lane")).toEqual("1");
  });

  test("adds running class when horse is running", () => {
    const horse = { color: "green", travelledDistance: 150, run: true };
    const wrapper = mount(Lane, {
      props: { horse, laneNo: 2 },
    });

    const horseDiv = wrapper.find(".horse");
    expect(horseDiv.classes()).toContain("running");
  });
});
