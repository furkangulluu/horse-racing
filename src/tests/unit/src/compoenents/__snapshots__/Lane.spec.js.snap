// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Lane.vue > is correctly rendered 1`] = `
"<div class="lane flex items-center" data-lane="1">
  <div class="horse overflow-hidden" style="color: rgb(0, 0, 0); margin-left: 0px;"> 🐎 </div>
</div>"
`;
